[Merged by Bors] - tortoise: describe tortoise implementation in README (#3031)
spacemeshos/go-spacemesh: tortoise/README.md

Tortoise is used to make decisions on the blocks that will be applied to the ledger. The protocol is asynchronous and makes progress based on the received data, without timing assumptions.

> ## Core concepts

Reviewer: i think it's worth adding 
- hdist
- zdist
- local opinions
- global opinions
- local threshold
- global threshold 
- live tortoise
- rerun tortoise
- sync tortoise
- verifying tortoise
- full tortoise
- self healing
in core concepts

Reviewer 2: Seconded - this document assumes a lot of prior knowledge and having way more context here would be helpful

#### Block

From the tortoise perspective it is important that the block has a unique ID and belongs to some layer.

> In the rest of the system block is a wrapper for transactions and rewards. So once tortoise decides on a block - related transactions and rewards will be applied to the ledger.

Reviewer 3: q: are we 100% sure that rewards must be part of a block's content? We have discussed this many times in the past but I'm not sure what's the latest consensus on this question.

Reviewer 4: a: Yes, we're 100% sure we need explicit rewards under UCB. The reason is that blocks under this model have no reference to the smeshers that provided input for them. This information is known at the time of creating the block but it's ephemeral, so can't be used for making local decisions (crediting some account).

The rewards in the block will not specify an explicit amount: when constructing the block we can't know the exact fees collected - the transactions must be executed to know for sure. Instead, the rewards will specify a weight per coinbase account and the full layer reward (incl. fees) will be distributed between the accounts by weight.

This approach will also make it easier to provide rewards for other roles than layer committee member (the ones that create ballots and proposals), like Hare participants and one day validators. For now, we'll only reward layer committee members, though.

#### Ballot

Tortoise counts votes from the syntactically valid ballots.

> For tortoise purposes every ballot must have a unique ID, reference to the base ballot ID, explicit votes (exceptions), weight and a beacon.

Reviewer: briefly describe what are in the exceptions?

> ###### Base ballot and exceptions

Reviewer 5: Although it's a header, it looks smaller than the text. Perhaps level 5 would be better?

Discussed in [Voting section](#voting).

###### Weight

> Weight of the ballot is computed from the weight of the activation transaction divided by the total number of eligible ballots. Total number of eligible ballots is computed based on the fraction of actiovation transaction compared to the weight from the active set that was recorded in reference ballot.

Reviewer 5: ```suggestion
Weight of the ballot is computed from the weight of the activation transaction divided by the total number of eligible ballots. Total number of eligible ballots is computed based on the fraction of activation transaction compared to the weight from the active set that was recorded in reference ballot.
```

###### Beacon

> Ballot with an inconsistent beacon will not be marked good and will not be counted immediatly. Discussed in [full](#votes-from-ballots-with-wrong-beacon) and [verifying](#good-ballots) sections.

Reviewer: maybe a short intro to beacon? 

`Beacon is a source of randomness and is calculated by the all smeshers per epoch. Ballots with a different beacon is either malicious or a result of a network split.`

> Ballot with an inconsistent beacon will not be marked good and will not be counted immediatly. Discussed in [full](#votes-from-ballots-with-wrong-beacon) and [verifying](#good-ballots) sections.

Reviewer 2: ```suggestion
Ballot with an inconsistent beacon will not be marked good and will not be counted immediately. Discussed in [full](#votes-from-ballots-with-wrong-beacon) and [verifying](#good-ballots) sections.
```

#### Expected weight and thresholds

> Expected weight is a total weight from the activation transactions. Unlike weight for the ballot we are using a sum of all activation transactions that target a specific epoch. Expected weight for the layer is fraction of the expected weight for the epoch.

Reviewer: > Expected weight for the layer is fraction of the expected weight for the epoch.

it's exactly total epoch weight divided by layers per epoch. 
also, the epoch weight is precise. it should not be termed as "expected weight for the epoch"

Reviewer 2: Probably worth taking a step back here and explaining the concept of weight and why it matters here

> 

Reviewer: thresholds are not mentioned. maybe a separate section for thresholds?

## Voting

> When voting protocol chooses base ballot and encodes the difference between local opinion and base ballot votes. Since we don't keep votes from genesis in memory the implementation has a shortcoming that it encodes difference only within some configurable window (we can refer to it as voting window). 

Reviewer: what is local opinions? i think there should be a separate section in core concepts about `local opinions` and `global opinions`

> In order to minimize amount of data that needs to be sent and stored, Tortoise selects ballot that has the least difference with a local opinion. 

Reviewer 2: How is this selection performed?

Types of votes:
- support

>     Support all blocks according to local opinion that are not supported by base ballot. Usually support is added for layers starting from base ballot layer up to last layer. In the event of healing tortoise may add support for blocks before base ballot layer.

Reviewer: what is last layer?

>     Support all blocks according to local opinion that are not supported by base ballot. Usually support is added for layers starting from base ballot layer up to last layer. In the event of healing tortoise may add support for blocks before base ballot layer.

Reviewer 2: Why would a ballot choose to support a block not supported by the base ballot?

- against
  

>     Against is added only if the base ballot supports the block and local opinion is against that block. 

Reviewer 2: Why/under what circumstances might this happen?

Author: this depends on the strategy for choosing base ballot. it might be more efficient to choose a ballot from later layer, and vote against blocks from previous layers. we are not doing this kind of optimizations

- abstain (undecided) 
    

>     Undecided is a valid vote only until hare is terminated for the layer (it may take more than one layer for hare to terminate, this is expressed as zdist).

Reviewer 2: Why is this an invalid vote in other circumstances?

> If ballot doesn't specify explicit vote for a block then it votes agains the block. This prevents malicious smeshers from retroactively reorganizing mesh by intentionally creating late ballots.

Reviewer: how does implicit against help prevent malicious reorg with late ballots? i thought this is about late blocks.

Author: yes thanks it is about creating late blocks

Reviewer 2: I think this explanation needs to be fleshed out further. How does "implicit against" prevent this attack?

> If ballot doesn't specify explicit vote for a block then it votes agains the block. This prevents malicious smeshers from retroactively reorganizing mesh by intentionally creating late ballots.

Reviewer 2: ```suggestion
If ballot doesn't specify explicit vote for a block then it votes against the block. This prevents malicious smeshers from retroactively reorganizing mesh by intentionally creating late ballots.
```

There are 4 distinct reasons to cast a vote:
- hare output
    
    tortoise uses output from the hare consensus for voting on blocks within hdist.

- validity/decisions

>     for layer before hdist the vote is casted according to the decision. 

Reviewer: why should be have two terms for the same thing? i'd prefer to say validity

>     for layer before hdist the vote is casted according to the decision. 

Reviewer: :s/casted/cast

>     for layer before hdist the vote is casted according to the decision. 

Reviewer 2: ```suggestion
    for layer before hdist the vote is cast according to the decision. 
```

>     for layer before hdist the vote is casted according to the decision. 

Reviewer 2: What decision? Why?

- local threshold

    in case when decision wasn't reached but margin from counted weight crosses local threshold tortoise will cast a vote according to the local threshold.

> - weak coin

Reviewer: how is weak coin generated?

Author: i don't remember details. would prefer not to look them up

>     otherwise tortoise votes according to the coin recorded in the last layer. if sufficient number of honest votes were casted this way then tortoise will either reach decision or atleast vote according to the local threshold in future layers. if honest nodes disagreed on the coin value then the voting will continue until they agree.

Reviewer 2: ```suggestion
    otherwise tortoise votes according to the coin recorded in the last layer. if sufficient number of honest votes were cast this way then tortoise will either reach decision or at least vote according to the local threshold in future layers. if honest nodes disagreed on the coin value then the voting will continue until they agree.
```

| 0xbb         | 20     | 1    | 1    | -1   | 1    |
| total        | -      | 30   | 10   | -30  | 30   |
| threshold    | -      | 20   | 20   | 20   | 20   |
| decision     | -      | 1    | 0    | -1   | 1    |

In the above example there are two ballots (0xaa and 0xbb). They vote for 4 blocks. Blocks 0x11 and 0x44 are valid and can be applied to the ledger, block 0x33 is invalid, and block 0x22 is still undecided.

#### Votes from ballots with wrong beacon

Because a smesher can select arbitrary beacon, there is an attack that will allow malicious smeshers to concentrate weight in a specific layer. If tortoise counts malicious weight immediatly it will make erroneous decisions (decide that a block is valid when in fact it is not). In order to prevent such attack tortoise will delay counting ballots with a wrong beacon until enough weight from honest smeshers was counted. 

Note that we can't discard ballots with wrong beacon completely as recovery from partition requires counting ballots from both sides. And they will likely have different beacons if it was a long partition, or occured at the time when tortoise beacon was computed.

#### Scaling issues

> Full tortoise complexity grows with the number of layers that are not finalized. 

Reviewer 2: Why/how? Computational, memory, messages, or some other form of complexity?

Author: mostly computational, for every new ballot we have to update all blocks in all undecided layers

> During live-tortoise this can be a problem only in extreme cases when recent layer fails to be finalized. Due to the excessive malicious voting (> 2/3 fraction of total weight) or insufficient voting from honest smeshers (due to the network problems).

Reviewer 2: What is meant by live-tortoise?

Author: live-tortoise is what we doing when it is not a rerun or sync

> However, during so called rerun (or sync) tortoise needs to count all votes. Otherwise there is always a risk that votes from latest layers changed decision that was made without counting them. With optimized full tortoise it might be practical to count votes when there are 2000 undecided layers, but not higher.

Reviewer: what does "count all votes" mean here? do you mean during rerun/sync we need to run full mode?

Author: in this case i meant count votes from all layers, not necessarily all votes from all layers, that depends on how we count. will try to rephrase

> However, during so called rerun (or sync) tortoise needs to count all votes. Otherwise there is always a risk that votes from latest layers changed decision that was made without counting them. With optimized full tortoise it might be practical to count votes when there are 2000 undecided layers, but not higher.

Reviewer 2: How is the number 2000 derived here?

In practice during rerun layer is expected to be finalized after tortoise counted votes from some limited number of layers, this is referred to as verification window in the codebase.  

## Verifying tortoise

> Verifying tortoise is meant as a solution for (scaling-issues)[#scaling-issues]of the full tortoise.

Reviewer 5: Not rendered as a link

> Verifying tortoise is meant as a solution for (scaling-issues)[#scaling-issues]of the full tortoise.

Reviewer 2: ```suggestion
Verifying tortoise is meant as a solution for (scaling-issues)[#scaling-issues] of the full tortoise.
```

| block | layer | opinion |
| ----- | ----- | ------- |
| 0x22  | 11    | -1      |
| 0x33  | 12    | 1       |
| 0x44  | 13    | 1       |

Important detail that abstain vote is consistent with any local opinion.Otherwise even a single abstain vote will force to switch tortoise into full mode.     

Whenever tortoise receives a ballot it checks if votes from this ballot are consistent with local table:

| block/ballot | layer | opinion | 0xaa | 0xbb |
| ------------ | ----- | ------- | ---- | ---- |
| 0x11         | 10    | 1       | 1    | 1    |
| 0x22         | 11    | -1      | 1    | -1   |
| 0x33         | 12    | 1       | 1    | 1    |
| 0x44         | 13    | 1       | -1   | 1    |

In this example only 0xbb ballot is consistent with local opinion. Consistency with a local opinion is a part of the definition of the good ballot, full definition is:
- ballot has a correct beacon
- ballot doesn't vote for block before base ballot layer
- ballot votes are consistent with local opinion 
- base ballot is good

> In the implementation we have a state named "can be good", the role of it's state will be clarified in [modes interaction](#modes-interaction) section. Ballot is marked "can be good" if first 3 condtions are satisfied, but not necessarily the last one.

Reviewer 2: ```suggestion
In the implementation we have a state named "can be good", the role of its state will be clarified in [modes interaction](#modes-interaction) section. Ballot is marked "can be good" if first 3 conditions are satisfied, but not necessarily the last one.
```

#### Vote counting in verifying mode

> Votes in verifying mode are counted in a batched form, unlike full tortoise. After "goodness" of the ballot is determinted tortoise filters out good ballots and sums up their weight. The result is a counted weight that is consistent with our local opinion. To get an accurate margin that can be compared with global threshold we assume that all uncounted ballots votes against our local opinion.

Reviewer: :s/determinted/determined

> Votes in verifying mode are counted in a batched form, unlike full tortoise. After "goodness" of the ballot is determinted tortoise filters out good ballots and sums up their weight. The result is a counted weight that is consistent with our local opinion. To get an accurate margin that can be compared with global threshold we assume that all uncounted ballots votes against our local opinion.

Reviewer: filter out good ballots means to remove good ballots. how about just `After "goodness" of the ballot is determined tortoise sums up the weight of good ballots`

Author: thanks, i guess i wanted to write filter good ballots, but your variant is better

## Modes interaction

> Verifying mode is safer as tortoise doesn't have an imposed limit on the amount of counted votes (or atleast imposed limit can be signicantly higher). Naturally we want to stay in verifying mode as long as tortoise can make progress.

Reviewer: :s/atleast/at least

> Verifying mode is safer as tortoise doesn't have an imposed limit on the amount of counted votes (or atleast imposed limit can be signicantly higher). Naturally we want to stay in verifying mode as long as tortoise can make progress.

Reviewer: what is this `imposed limit on the amount of counted votes`?

Author: that's about verification window. need to link it somehow

> Verifying mode is safer as tortoise doesn't have an imposed limit on the amount of counted votes (or atleast imposed limit can be signicantly higher). Naturally we want to stay in verifying mode as long as tortoise can make progress.

Reviewer 2: What is meant by "safer" here? Why is it safer?

Author: because we don't have a limit on number of layers that we will count before the target layer is verified.

Author: it is actually described in the rest of the sentance

#### When to switch into full mode?

> Tortoise certainly has to switch into full mode if there is a layer that wasn't finalized past hdist layers. In such case it needs to count previously uncounted ballots so that real margin can be compared with local threshold, see [voting](#voting). Without counting all votes we can get only pessimistic margin, which works for the purposes of the verifying tortoise, but may not work for self-healing purposes (e.g. instead of voting according to local threshold sign tortoise will vote according to weak coin).

Reviewer: > real margin can be compared with local threshold

do you mean global threshold?

Author: no, i meant that we if we have "real" margin it may cross local threshold, when "conservative" margin may not. so whenever we vote on undecided layers past hdist we want to compute "real" margin.

"real" is for margin computed using full tortoise. "conservative" is for - verifying.

> Tortoise certainly has to switch into full mode if there is a layer that wasn't finalized past hdist layers. In such case it needs to count previously uncounted ballots so that real margin can be compared with local threshold, see [voting](#voting). Without counting all votes we can get only pessimistic margin, which works for the purposes of the verifying tortoise, but may not work for self-healing purposes (e.g. instead of voting according to local threshold sign tortoise will vote according to weak coin).

Reviewer: > but may not work for self-healing purposes

this is the first time self-healing is mentioned in this doc. i think it needs a brief intro in core concept too.

During rerun we expect verifying tortoise to make progress after counting `verifying-mode-verification-window` layers. This window can be as large as memory will allow us, computational complexity doesn't increase with the window. However if tortoise didn't make progress after counting votes within this window it will have to switch into full mode during rerun. 

#### When to switch into verifying mode?

It is safe to make an attempt at any point. However we need a sufficient number of ballots that vote consistently with each other in order for verifying tortoise to work. 

> In the implementation, we make an attempt after full tortoise made progress and changed local opinion on layers that were undecided according to the verifying tortoise. In the layer that was just finalized by full tortoise we filter ballots that ["can be good"](#good-ballots) with base ballots in the same state. After they are marked good we restart verifying tortoise and if same layer it made the same progress as full tortoise we switch the mode.

Reviewer: :s/we filter ballots/we find ballots

| block/ballot | local opinion | 0xaa | 0xbb | 0xcc | 0xdd | 0xee | 0xff |
| ------------ | ------------- | ---- | ---- | ---- | ---- | ---- | ---- |
| base         | -             | -    | -    | 0xaa | 0xaa | 0xcc | 0xcc |
| layer        | -             | 10   | 10   | 11   | 11   | 12   | 12   |
| 0x11         | 1             | -1   | -1   | 1    | -1   | 1    | 1    |
| 0x22         | 1             | -1   | -1   | 1    | -1   | 1    | 1    |
| 0x33         | -1            | -    | -    | -1   | -1   | -1   | -1   |
| 0x44         | -1            | -    | -    | -1   | -1   | -1   | -1   |
| 0x55         | 1             | -    | -    | -    | -    | 1    | 1    |
| 0x66         | -1            | -    | -    | -    | -    | -1   | -1   |

> In this example ballots starting from layer 10 are voting for some old layer (for example 9). Ballots from layer 10 disagree with out local opinion and will be marked bad. Ballot 0xcc from layer 10 and ballots 0xee and 0xff will be marked "can be good" because exceptions from those ballots are consistent with local opinion.

Reviewer 5: ```suggestion
In this example ballots starting from layer 10 are voting for some old layer (for example 9). Ballots from layer 10 disagree with our local opinion and will be marked bad. Ballot 0xcc from layer 10 and ballots 0xee and 0xff will be marked "can be good" because exceptions from those ballots are consistent with local opinion.
```
